src/tokens/selectors.ts

```diff
@@ -42,21 +43,97 @@ const DEFAULT_MEMOIZE_MAX_SIZE = 10
 
 function isNetworkIdList(networkIds: any): networkIds is NetworkId[] {
   return (
-    networkIds.constructor === Array &&
+    Array.isArray(networkIds) &&
     networkIds.every((networkId) => Object.values(NetworkId).includes(networkId))
   )
 }
 export const tokenFetchLoadingSelector = (state: RootState) => state.tokens.loading
 export const tokenFetchErrorSelector = (state: RootState) => state.tokens.error
 
+// Note: not importing from 'src/positions/selectors' to avoid circular dependency
+// TODO: address circular dependency
+const positionsSelector = (state: RootState) => state.positions.positions
+const positionsFetchedAtSelector = (state: RootState) => state.positions.positionsFetchedAt
+
+const positionTokensSelector = createSelector([positionsSelector], (positions) => {
+  const positionTokens: Record<string, Token> = {}
+
+  function visitToken(token: Token, isTopLevelAppToken: boolean) {
+    const tokenId = token.tokenId
+    if (!positionTokens[tokenId] || isTopLevelAppToken) {
+      positionTokens[tokenId] = {
+        ...token,
+        // Only keep the balance if it's a top level app-token
+        // Otherwise it doesn't represent a real user balance
+        balance: isTopLevelAppToken ? token.balance : '0',
+      }
+    }
+    if (token.type === 'app-token') {
+      for (const childToken of token.tokens) {
+        visitToken(childToken, false)
+      }
+    }
+  }
+
+  for (const position of positions) {
+    if (position.type === 'app-token') {
+      visitToken(position, true)
+    } else {
+      for (const token of position.tokens) {
+        visitToken(token, false)
+      }
+    }
+  }
+  return positionTokens
+})
+
 export const tokensByIdSelector = createSelector(
   [
     (state: RootState) => state.tokens.tokenBalances,
+    positionTokensSelector,
+    positionsFetchedAtSelector,
     (_state: RootState, networkIds: NetworkId[]) => networkIds,
   ],
-  (storedBalances, networkIds) => {
+  (storedBalances, positionTokens, positionsFetchedAt, networkIds) => {
+    const allStoredBalances = { ...storedBalances }
+
+    // Enrich with position tokens
+    // This allows us to have priceUsd and balance for tokens
+    // decomposed via positions
+    for (const positionToken of Object.values(positionTokens)) {
+      if (!networkIds.includes(positionToken.networkId)) {
+        continue
+      }
+      const tokenId = positionToken.tokenId
+      const existingToken = allStoredBalances[tokenId]
+      const priceUsd = positionToken.priceUsd != '0' ? positionToken.priceUsd : undefined
+      if (!existingToken) {
```

Comment:
adding all position tokens here slightly makes me nervous 😅  since its a backwards incompatible change and its on the consumers to filter out these tokens as required. Right now I see there are filters in some selectors, but some others don't seem to include it, like cash in / cash out token selectors, tokens with balance selector etc. Maybe these are fine for now, but wonder if down the line we'll run into some bugs, as its not super obvious from the name of the selector on whether it includes or excludes position tokens. Couple of ideas (not fully thought through):
- Make this change backwards compatible, e.g., don't include these by default and only include if requested by consumers. Maybe this is hard to do with selector parameters
- Have some naming pattern that makes it obvious that the selector includes or excludes position tokens.

What do you think?

Reply:
Great feedback, I was going back and forth about this.
I guess there are too parts here.

1. Injecting the $ price everywhere if not already present. I think this is the right place for this.
2. Adding all tokens from positions. I added it here for now so all places using `tokenId` could access position tokens (for instance `TokenDisplay` working out of the box with `tokenId` for positions). But the drawback is we don't want them in all selectors down the line. That's why I resorted to adding `isFromPosition`. Which isn't ideal like you mentioned. I tried to be careful, but it's possible it can cause subtle bugs. On a side note, Cash In and Cash Out token selectors don't need more filtering because they already check if `isCashInEligible` / `isCashOutEligible` is present. But it's possible I missed something else or that it will cause confusion in the future.

Anyway to be more on the safe side, I'll take a stab at making it a backward compatible change.

Reply:
Changed in https://github.com/valora-inc/wallet/pull/5659/commits/cb86d4322ea709e89bb828fa2f9ba4e44958ec37

Let me know what you think @satish-ravi 🙏

Reply:
This looks great! Thanks!